File: hasad/backend/Hasad.Domain/Entities/DamageReport.cs
using Hasad.Domain.Common;

namespace Hasad.Domain.Entities;

public class DamageReport : ISoftDelete
{
    public Guid Id { get; set; }
    public Guid ClientId { get; set; }

    public string ReportNumber { get; set; } = string.Empty;
    public string PermanentFormNumber { get; set; } = string.Empty;
    public string TemporaryFormNumber { get; set; } = string.Empty;

    public Guid FarmId { get; set; }
    public Farm? Farm { get; set; }

    public Guid FarmerId { get; set; }
    public Guid GovernorateId { get; set; }
    public Guid DirectorateId { get; set; }
    public Guid LocalityId { get; set; }
    public int DamageYear { get; set; }

    public DateTime DamageDate { get; set; }
    public DateTime DocumentationDate { get; set; }

    public int AgriculturalSectorId { get; set; }
    public AgriculturalSector? AgriculturalSector { get; set; }

    public int DamageCauseCategoryId { get; set; }
    public DamageCauseCategory? DamageCauseCategory { get; set; }

    public int DamageCauseId { get; set; }
    public DamageCause? DamageCause { get; set; }

    public string StatusId { get; set; } = "PendingTechnicalVerification";

    public decimal TotalDamage { get; set; }

    public string Notes { get; set; } = string.Empty;

    public string CreatedBy { get; set; } = string.Empty;
    public DateTime CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }

    public byte[] RowVersion { get; set; } = Array.Empty<byte>();

    public bool IsDeleted { get; set; }
    public DateTime? DeletedAt { get; set; }
    public string? DeletedBy { get; set; }

    public ICollection<DamageItem> Items { get; set; } = new List<DamageItem>();
    public ICollection<DamageReportAttachment> Attachments { get; set; } = new List<DamageReportAttachment>();
}

File: hasad/mobile/lib/features/damage_reports/domain/models/damage_report.dart
// ignore_for_file: invalid_annotation_target
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:mobile/features/damage_reports/domain/models/damage_item.dart';
import 'package:mobile/features/damage_reports/domain/models/damage_report_status.dart';

part 'damage_report.freezed.dart';
part 'damage_report.g.dart';

@freezed
class DamageReport with _$DamageReport {
  const factory DamageReport({
    @JsonKey(name: 'clientId') @Default('') String id, // ClientId
    @JsonKey(name: 'id') String? serverId,
    @Default('') String reportNumber,
    @Default('') String permanentFormNumber,
    @Default('') String temporaryFormNumber,
    @Default(0) int damageYear,
    @Default('') String farmId,
    @Default('') String farmerId,
    DateTime? damageDate,
    DateTime? documentationDate,
    @Default(0) int agriculturalSectorId,
    @Default(0) int damageCauseCategoryId,
    @Default(0) int damageCauseId,
    @Default('') String governorateId,
    @Default('') String directorateId,
    @Default('') String localityId,
    @Default(DamageReportStatus.pendingTechnicalVerification) String statusId,
    @JsonKey(name: 'totalDamage') @Default(0.0) double totalDamage,
    @Default('') String notes,
    @Default('') String createdBy,
    @Default('') String rowVersion,
    @Default([]) List<DamageItem> items,
    @Default('completed') String syncStatus,
    String? lastSyncError,
    DateTime? updatedAt,
    bool? isDeleted,
    DateTime? deletedAt,
    String? deletedBy,
  }) = _DamageReport;

  factory DamageReport.fromJson(Map<String, dynamic> json) =>
      _$DamageReportFromJson(json);
}

extension DamageReportWorkflowX on DamageReport {
  bool get isHeaderSynced => serverId != null && serverId!.isNotEmpty && reportNumber.isNotEmpty;

  bool get areItemsSynced => items.every((item) => item.syncStatus == 'completed');

  bool get hasItems => items.isNotEmpty;

  bool get isReadyForReview {
    return isHeaderSynced &&
        hasItems &&
        areItemsSynced &&
        statusId != DamageReportStatus.completed &&
        statusId != 'Submitted'; // Check for legacy status if any
  }

  String get workflowStateKey {
    if (statusId == 'Submitted' || statusId == DamageReportStatus.techReview) return 'status_TechReview';
    if (statusId == DamageReportStatus.completed) return 'status_Completed';

    if (!isHeaderSynced) {
      return syncStatus == 'failed' ? 'workflowState_HeaderSyncFailed' : 'workflowState_DraftHeader';
    }

    if (!hasItems) return 'workflowState_HeaderSynced';

    if (!areItemsSynced) return 'workflowState_AssessmentPendingSync';

    if (isReadyForReview) return 'workflowState_ReadyForReview';

    return 'workflowState_AssessmentInProgress';
  }
}
